src/osd/PG.cc

```diff
@@ -4207,24 +4201,44 @@ void PG::scrub(epoch_t queued, ThreadPool::TPHandle &handle)
 {
   if (cct->_conf->osd_scrub_sleep > 0 &&
       (scrubber.state == PG::Scrubber::NEW_CHUNK ||
-       scrubber.state == PG::Scrubber::INACTIVE) && scrubber.needs_sleep) {
+       scrubber.state == PG::Scrubber::INACTIVE) &&
+       scrubber.needs_sleep) {
     ceph_assert(!scrubber.sleeping);
     dout(20) << __func__ << " state is INACTIVE|NEW_CHUNK, sleeping" << dendl;
+
     // Do an async sleep so we don't block the op queue
-    auto scrub_requeue_callback = new FunctionContext([this](int r) {
-      lock();
-      scrubber.sleeping = false;
-      scrubber.needs_sleep = false;
-      dout(20) << __func__ << " slept for "
-               << ceph_clock_now() - scrubber.sleep_start
-               << ", re-queuing scrub" << dendl;
-      scrub_queued = false;
-      requeue_scrub();
-      scrubber.sleep_start = utime_t();
-      unlock();
-    });
-    Mutex::Locker l(scrub_sleep_lock);
-    scrub_sleep_timer.add_event_after(cct->_conf->osd_scrub_sleep, scrub_requeue_callback);
+    OSDService *myosdservice = osd;
+    spg_t pgid = get_pgid();
+    auto scrub_requeue_callback =
+        new FunctionContext([myosdservice, pgid](int r) {
+          PG *pg = nullptr;
+          {
+            RWLock::RLocker l(myosdservice->osd->pg_map_lock);
```

Comment:
you may as well make a wrapper for OSD::_lookup_lock_pg in OSDService.

This one has the problem of potentially dereferencing an invalid pg still, since it drops the pg_map_lock before taking the pg->lock(), leaving a window with no locks held during which the pg could be removed.